### DISCUSSION TITLE
Suggest me the best approach for whiteboard

### DISCUSSION THREAD

**Discussion by Tonku** [pocketbase/pocketbase#5779] [Q&A]
# Best Approach for Storing and Managing Large JSON Documents in a Collaborative Whiteboard Application

## Description

Hello,

I'm currently building a collaborative whiteboard application where the document data (e.g., design elements, images, text) is represented as a deeply nested JSON structure. The JSON document may grow to be around 1-2 MB in size.

After reviewing discussions about JSON field sizes, I noticed that @ganigeorgiev recommended considering anything over 5KB as large, as it may impact performance. Given this guidance, I’m a bit uncertain about the best approach to handle this data efficiently.

## Requirements and Current Approach

- **Frequent Edits**: Users will frequently interact with and modify the document, requiring regular updates to the JSON.
- **Incremental Updates**: Rather than sending the entire JSON document from the client with each change, I plan to use a JSON patching approach. Specifically, I’m considering applying patches using [json-patch](https://github.com/evanphx/json-patch) in a custom Go hook to update only the modified parts of the JSON.

## Questions

1. **Performance Concerns**: Given that PocketBase considers large JSON fields to potentially impact performance, will storing several large JSON documents as records in my database slow down my server as the number of records grows?
   
2. **Alternative Approaches**: Would it be more efficient to treat the document as a file (e.g., JSON or binary format) and store it in an external bucket, such as S3? If so, how would this impact my ability to update the JSON document efficiently?

3. **Future Collaboration Requirements**: Although collaborative editing is not currently planned, it may be added in the future. Would either approach be better suited to support real-time, multi-user interactions down the line?

Any guidance on the best way to structure and store these large documents for efficient performance and scalability would be greatly appreciated.

Thank you!

**Comment by Tonku** [comment]
I think I can use a temporary local json file and apply json patches on it as the user edits and finally update the db one the editing is completed

**Comment by thangngoc89** [comment]
Instead of reinventing the wheel, you might want to look at CRDT, which automerge is one of the JS implementations. Trying to make this syncing robustness is madness if you start from scratch
